### ISSUE TITLE
Pulling should be done asynchronously

### ISSUE THREAD

**Issue by apple502j** [transifex/cli#5]
Currently languages are pulled one by one, which makes this process significantly slower (5-10 minutes to pull approx. 120 files). Multiple pulls should be processed at the same time.

**Comment by kovidgoyal** [comment]
Better than just asynchronously, you should send the hashes of all po files in a single request and have the server reply with a list of updated files and then download *those* in parallel.

Going from the old client, which had --parallel to the new client has changed the download time for my project with tx pull -a from a few minutes to a few hours.

**Comment by dan-leanix** [comment]
The new client is damn slow, and it is advertised as faster then the previous client: https://github.com/transifex/cli#differences-with-the-previous-version

Our pipeline step for pulling translations increased from 1 minute to 10 minutes. Factor 10x slower.

Looking at the following statements, the performance of the new client needs to be drastically improved.

> We released [a stable version of a new Transifex Client](https://github.com/transifex/cli/releases), compatible with the [API 3.0](https://transifex.github.io/openapi/), and offered as single executable.
> 
> Please start using the new Transifex CLI [here](https://github.com/transifex/cli), since this software is considered deprecated (as of January 2022) and will sunset this year.

> We are deprecating the API versions namely API v2 and v2.5, to make way for the more usable and maintainable [API v3](https://transifex.github.io/openapi/index.html#section/Introduction).
> 
> As of April 7th, 2022 the API versions 2 and 2.5 will no longer be operational and relevant requests will begin to fail.

**Comment by mathaeus** [comment]
We also experienced an increase of duration by a factor of more than 10 (I think it was factor of 13 or so).
I was in contact with Transifex about the issue and they promised to do something about it, but also mentioned that this is of course something that takes a while.

@codegaze since the deadline is approaching, but obviously the new client is not usable for many of us, can you make any official statement about if the deadline for dropping support for API v2 will be moved?

**Comment by gmatikas** [comment]
We are also experiencing an increase of duration by a factor of a lot more than 10 using the new CLI client compared to the python one.

**Comment by nelefth** [comment]
Thank you all very much for your active participation and contribution to this discussion. We completely understand the importance of addressing this issue. Our team is currently evaluating the situation. We will provide you with an update early next week. 
We truly appreciate your cooperation and patience in this matter.

**Comment by nelefth** [comment]
Hi all,

I would like to follow up on this and share 2 important updates :) 

- The sunset of the deprecated API 2.0/2.5 and Transifex client, originally scheduled for April 7, 2022, has been postponed to Nov 30, 2022. [Here](https://community.transifex.com/t/postponing-api-2-0-2-5-and-transifex-client-sunset-date/2759) you can find our official announcement. 
- Regarding the initial request, we are currently working on reducing the requests needed to pull/push files as well as extending Go CLI capabilities by including the parallel actions feature.

For any questions, please do not hesitate to reach out to us.

**Comment by Swyter** [comment]
Even easier; have a server-side UTC timestamp of the last time the resource was modified, and check if it's newer than the local file's last-modified date. Careful with timezones.

But, yeah. Downloading all the `resource_count * language_count` files one by one takes a long time, especially when done unnecessarily. The previous version detected unchanged resources and avoided the download altogether.

Downloading prebuilt versions of the old client was also hard because the CI artifacts have timed out by now. So oeople may be grabbing random `tx.exe` binaries from the Internet/other projects. Having a GitHub release with the final official version of the previous client would come in handy.

PS: Having an option for a less chatty download progress (no spinner/animations/colors) in the new client would also come in handy. It ends up printing a lot of lines in the CI log.

**Comment by codegaze** [comment]
Hi all!

 I’d like to give you an update on this front. In the new version of the CLI ([v1.2.0](https://github.com/transifex/cli/releases/tag/v1.2.0)) we introduced the concurrent functionality for `pull` and `push` which should resolve the issues mentioned in this thread.
 
By default we use 5 workers, but you can always set a custom number by using the `—workers` flag.  These are all the changes of the latest version.  * Introduce parallel push/pull 

* Removed PTerm in favour of simple color package to enable real time updates and reduce noise in CIs. 
* Performance fix for `--use-git-timestamps` when history is big. 
* Introduce "wait" mechanism on API throttling.
* Reduced the number of calls to the API.

Feel free to try it out and reach out for any question or feedback!
